homeassistant/components/zabbix/__init__.py

```diff
@@ -46,13 +90,208 @@ def setup(hass, config):
+                and state.domain not in whitelist_d
+            ):
+                return
+
+            _state_as_value = float(state.state)
+            floats[entity_id] = _state_as_value
+        except ValueError:
+            try:
+                _state_as_value = float(state_helper.state_as_number(state))
+                floats[entity_id] = _state_as_value
+            except ValueError:
+                strings[entity_id] = state.state
+
+        for key, value in state.attributes.items():
+            # For each value we try to cast it as float
+            # But if we can not do it we store the value
+            # as string
+            attribute_id = f"{entity_id}/{key}"
+            try:
+                float_value = float(value)
+                if math.isfinite(float_value):
+                    floats[attribute_id] = float_value
+                else:
+                    strings[attribute_id] = str(value)
+            except (ValueError, TypeError):
+                strings[attribute_id] = str(value)
+
+        metrics = []
+        float_keys_count = len(float_keys)
+        float_keys.update(floats.keys())
+        if len(float_keys) != float_keys_count:
+            floats_discovery = []
+            for float_key in float_keys:
+                floats_discovery.append({"{#KEY}": float_key})
+            metric = ZabbixMetric(
+                publish_states_host,
+                "homeassistant.floats_discovery",
+                json.dumps(floats_discovery),
+            )
+            metrics.append(metric)
+        for key, value in floats.items():
+            metric = ZabbixMetric(
+                publish_states_host, f"homeassistant.float[{key}]", value
+            )
+            metrics.append(metric)
+
+        string_keys.update(strings.keys())
+        return metrics
+
+    def shutdown(event):
+        """Shut down the thread."""
+        instance.queue.put(None)
+        instance.join()
+
+    if publish_states_host:
+        zabbix_sender = ZabbixSender(zabbix_server=conf[CONF_HOST])
+        _LOGGER.info("Initialized Zabbix sender")
+        instance = ZabbixThread(hass, zabbix_sender, event_to_metrics, max_tries)
+        instance.start()
+        _LOGGER.info("Started ZabbixThread")
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, shutdown)
+
     return True
+
+
+class ZabbixThread(threading.Thread):
+    """A threaded event handler class."""
+
+    def __init__(self, hass, zabbix_sender, event_to_metrics, max_tries):
+        """Initialize the listener."""
+        threading.Thread.__init__(self, name="Zabbix")
+        self.queue = queue.Queue()
+        self.zabbix_sender = zabbix_sender
+        self.event_to_metrics = event_to_metrics
+        self.max_tries = max_tries
+        self.write_errors = 0
+        self.shutdown = False
+        self.float_keys = set()
+        self.string_keys = set()
+
+        hass.bus.listen(EVENT_STATE_CHANGED, self._event_listener)
```

Comment:
We try to avoid side effects in init method. Use a dedicated setup method or move this out of the class.

Reply:
I added a setup method to ServerThread which subscribes to both events and starts the thread.